src/river.py

```diff
@@ -55,14 +63,22 @@ def end_time_step(self) -> None:
 interface.initialize_data()
 
 while interface.is_coupling_ongoing():
+    if interface.is_action_required(precice.action_write_iteration_checkpoint()):
+        river.save_state()
+        interface.mark_action_fulfilled(precice.action_write_iteration_checkpoint())
     dt = min(solver_dt, precice_dt)
 
     flux = interface.read_scalar_data(flux_id, vertex_id)
     river.solve(dt, flux)
     interface.write_scalar_data(height_id, vertex_id, river.height)
 
     precice_dt = interface.advance(dt)
-    river.end_time_step()
+
+    if interface.is_action_required(precice.action_read_iteration_checkpoint()):
+        river.load_state()
+        interface.mark_action_fulfilled(precice.action_read_iteration_checkpoint())
+    else:
```

Comment:
With this simple check, subcycling/multirate doesn't work properly. `end_time_step()` is called in two different cases:
- in each iteration for "intermediate" time steps
- only in the final iteration for the final time step of a time window

Ideally, I would like to save model output only in the final iteration for every time step. Maybe check https://precice.org/couple-your-code-time-step-sizes.html for this?

Reply:
Here's an example of the issue if `River` subcycles `Groundwater` (`N=40` in river.py):
![image](https://github.com/valentinaschueller/richards-swe-coupling/assets/64773821/48bce872-3c12-409a-a32e-da7ed30a5fac)

Reply:
I'll take care of this at a later point (see #15)